File: rust/src/leetcode/Q238_product_of_array_except_self.rs
/*
 * @lc app=leetcode id=238 lang=rust
 *
 * [238] Product of Array Except Self
 */

// @lc code=start
impl Solution {
    pub fn product_except_self(nums: Vec<i32>) -> Vec<i32> {
        let mut product: i32 = 1;
        let mut prefix: Vec<i32> = vec![1; nums.len()];

        for i in 0..nums.len() {
            if i != 0 {
                prefix[i] = product;
            }

            product *= nums[i];
        }

        product = 1;
        let mut suffix: Vec<i32> = vec![1; nums.len()];

        for i in (0..nums.len()).rev() {
            if i != nums.len() - 1 {
                suffix[i] = product;
            }

            product *= nums[i];
        }

        let mut result: Vec<i32> = vec![1; nums.len()];

        for i in 0..nums.len() {
            result[i] = prefix[i] * suffix[i]
        }

        result
    }
}
// @lc code=end

File: python/Q238_product_of_array_except_self.py
#
# @lc app=leetcode id=238 lang=python3
#
# [238] Product of Array Except Self
#


# @lc code=start
class Solution:
    def in_place(self, nums: list[int]) -> list[int]:
        length: int = len(nums)
        result: list[int] = [1] * length

        number: int = 1
        for i in range(1, length):
            number *= nums[i - 1]
            result[i] *= number

        number = 1
        for i in range(length - 2, -1, -1):
            number *= nums[i + 1]
            result[i] *= number

        return result

    def prefix_and_suffix(self, nums: list[int]) -> list[int]:
        prefix: list[int] = [1] * len(nums)

        for i in range(1, len(nums)):
            prefix[i] = prefix[i - 1] * nums[i - 1]

        suffix: list[int] = [1] * len(nums)

        for i in range(len(nums) - 2, -1, -1):
            suffix[i] = suffix[i + 1] * nums[i + 1]

        result: list[int] = [prefix[i] * suffix[i] for i in range(len(nums))]

        return result

    def productExceptSelf(self, nums: list[int]) -> list[int]:
        return self.in_place(nums)


# @lc code=end
